### PR TITLE
Relax Floating Point Number Tolerance for PCA test

### PR DESCRIPTION
**Pull request by yc7z**
# PR Type
[Fix]

Relaxed the floating point number tolerance used in `test_pca.py`.

# Tests Added

Updated `test_pca.py`

### CODE REVIEW

**Comment by lotif** [comment]
Tests still fail for me :(
```
============================================================================================================= FAILURES =============================================================================================================
___________________________________________________________________________________________________ test_full_svd_low_rank_data ____________________________________________________________________________________________________

    def test_full_svd_low_rank_data() -> None:
        pca_module = PcaModule(low_rank=False, full_svd=True)
        X = create_low_rank_data()
        principal_components, singular_values = pca_module(X, center_data=True)
        assert principal_components.size(1) == data_dimension and len(singular_values) == data_dimension
        # Since X is low-rank, it is expected that only the first small_rank singular values should be nonzero.
        # The remaining singular values should be (nearly) zero.
>       assert torch.allclose(singular_values[small_rank:], torch.zeros(data_dimension - small_rank), atol=1e-5)
E       assert False
E        +  where False = <built-in method allclose of type object at 0x1083ee330>(tensor([9.5387e-05, 8.5341e-05, 8.1460e-05, 7.6568e-05, 6.7481e-05, 6.0333e-05,\n        5.0726e-05, 4.3949e-05, 3.8518...6, 1.7162e-06, 1.7162e-06, 1.7162e-06,\n        1.7162e-06, 1.7162e-06, 1.7162e-06, 1.7162e-06, 1.7162e-06, 1.7162e-06]), tensor([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n        0., 0....., 0., 0., 0.,\n        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]), atol=1e-05)
E        +    where <built-in method allclose of type object at 0x1083ee330> = torch.allclose
E        +    and   tensor([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n        0., 0....., 0., 0., 0.,\n        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]) = <built-in method zeros of type object at 0x1083ee330>((256 - 16))
E        +      where <built-in method zeros of type object at 0x1083ee330> = torch.zeros

tests/models/test_pca.py:83: AssertionError
------------------------------------------------------------------------------------------------------- Captured stderr call -------------------------------------------------------------------------------------------------------
INFO flwr 2024-02-01 14:04:36,233 | pca.py:91 | Performing full SVD on data matrix.
-------------------------------------------------------------------------------------------------------- Captured log call ---------------------------------------------------------------------------------------------------------
INFO     flwr:pca.py:91 Performing full SVD on data matrix.
__________________________________________________________________________________________________ test_reduced_svd_low_rank_data __________________________________________________________________________________________________

    def test_reduced_svd_low_rank_data() -> None:
        pca_module = PcaModule(low_rank=False, full_svd=False)
        X = create_low_rank_data()
        principal_components, singular_values = pca_module(X, center_data=True)
        assert principal_components.size(1) == data_dimension and len(singular_values) == data_dimension
>       assert torch.allclose(singular_values[small_rank:], torch.zeros(data_dimension - small_rank), atol=1e-5)
E       assert False
E        +  where False = <built-in method allclose of type object at 0x1083ee330>(tensor([8.9834e-05, 8.7510e-05, 8.2494e-05, 7.9508e-05, 6.9625e-05, 6.5962e-05,\n        5.8876e-05, 5.5998e-05, 4.4957...6, 1.7008e-06, 1.7008e-06, 1.7008e-06,\n        1.7008e-06, 1.7008e-06, 1.7008e-06, 1.7008e-06, 1.7008e-06, 1.7008e-06]), tensor([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n        0., 0....., 0., 0., 0.,\n        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]), atol=1e-05)
E        +    where <built-in method allclose of type object at 0x1083ee330> = torch.allclose
E        +    and   tensor([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n        0., 0....., 0., 0., 0.,\n        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]) = <built-in method zeros of type object at 0x1083ee330>((256 - 16))
E        +      where <built-in method zeros of type object at 0x1083ee330> = torch.zeros

tests/models/test_pca.py:108: AssertionError
------------------------------------------------------------------------------------------------------- Captured stderr call -------------------------------------------------------------------------------------------------------
INFO flwr 2024-02-01 14:04:36,447 | pca.py:93 | Performing reduced SVD on data matrix.
-------------------------------------------------------------------------------------------------------- Captured log call ---------------------------------------------------------------------------------------------------------
INFO     flwr:pca.py:93 Performing reduced SVD on data matrix.

```

**Comment by yc7z** [comment]
> Tests still fail for me :(
> 
> ```
> ============================================================================================================= FAILURES =============================================================================================================
> ___________________________________________________________________________________________________ test_full_svd_low_rank_data ____________________________________________________________________________________________________
> 
>     def test_full_svd_low_rank_data() -> None:
>         pca_module = PcaModule(low_rank=False, full_svd=True)
>         X = create_low_rank_data()
>         principal_components, singular_values = pca_module(X, center_data=True)
>         assert principal_components.size(1) == data_dimension and len(singular_values) == data_dimension
>         # Since X is low-rank, it is expected that only the first small_rank singular values should be nonzero.
>         # The remaining singular values should be (nearly) zero.
> >       assert torch.allclose(singular_values[small_rank:], torch.zeros(data_dimension - small_rank), atol=1e-5)
> E       assert False
> E        +  where False = <built-in method allclose of type object at 0x1083ee330>(tensor([9.5387e-05, 8.5341e-05, 8.1460e-05, 7.6568e-05, 6.7481e-05, 6.0333e-05,\n        5.0726e-05, 4.3949e-05, 3.8518...6, 1.7162e-06, 1.7162e-06, 1.7162e-06,\n        1.7162e-06, 1.7162e-06, 1.7162e-06, 1.7162e-06, 1.7162e-06, 1.7162e-06]), tensor([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n        0., 0....., 0., 0., 0.,\n        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]), atol=1e-05)
> E        +    where <built-in method allclose of type object at 0x1083ee330> = torch.allclose
> E        +    and   tensor([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n        0., 0....., 0., 0., 0.,\n        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]) = <built-in method zeros of type object at 0x1083ee330>((256 - 16))
> E        +      where <built-in method zeros of type object at 0x1083ee330> = torch.zeros
> 
> tests/models/test_pca.py:83: AssertionError
> ------------------------------------------------------------------------------------------------------- Captured stderr call -------------------------------------------------------------------------------------------------------
> INFO flwr 2024-02-01 14:04:36,233 | pca.py:91 | Performing full SVD on data matrix.
> -------------------------------------------------------------------------------------------------------- Captured log call ---------------------------------------------------------------------------------------------------------
> INFO     flwr:pca.py:91 Performing full SVD on data matrix.
> __________________________________________________________________________________________________ test_reduced_svd_low_rank_data __________________________________________________________________________________________________
> 
>     def test_reduced_svd_low_rank_data() -> None:
>         pca_module = PcaModule(low_rank=False, full_svd=False)
>         X = create_low_rank_data()
>         principal_components, singular_values = pca_module(X, center_data=True)
>         assert principal_components.size(1) == data_dimension and len(singular_values) == data_dimension
> >       assert torch.allclose(singular_values[small_rank:], torch.zeros(data_dimension - small_rank), atol=1e-5)
> E       assert False
> E        +  where False = <built-in method allclose of type object at 0x1083ee330>(tensor([8.9834e-05, 8.7510e-05, 8.2494e-05, 7.9508e-05, 6.9625e-05, 6.5962e-05,\n        5.8876e-05, 5.5998e-05, 4.4957...6, 1.7008e-06, 1.7008e-06, 1.7008e-06,\n        1.7008e-06, 1.7008e-06, 1.7008e-06, 1.7008e-06, 1.7008e-06, 1.7008e-06]), tensor([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n        0., 0....., 0., 0., 0.,\n        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]), atol=1e-05)
> E        +    where <built-in method allclose of type object at 0x1083ee330> = torch.allclose
> E        +    and   tensor([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n        0., 0....., 0., 0., 0.,\n        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]) = <built-in method zeros of type object at 0x1083ee330>((256 - 16))
> E        +      where <built-in method zeros of type object at 0x1083ee330> = torch.zeros
> 
> tests/models/test_pca.py:108: AssertionError
> ------------------------------------------------------------------------------------------------------- Captured stderr call -------------------------------------------------------------------------------------------------------
> INFO flwr 2024-02-01 14:04:36,447 | pca.py:93 | Performing reduced SVD on data matrix.
> -------------------------------------------------------------------------------------------------------- Captured log call ---------------------------------------------------------------------------------------------------------
> INFO     flwr:pca.py:93 Performing reduced SVD on data matrix.
> ```

My bad I misread your error message. It should work now

**Comment by lotif** [comment]
it works now, thanks for fixing!!!

File: tests/models/test_pca.py [hunk 1]
```diff
@@ -9,20 +11,21 @@
 N_small = 128
 seed = 83
 small_rank = 16
-set_all_random_seeds(seed)
```
Comment (lotif) [inline tests/models/test_pca.py:12]: I think it makes sense to keep it here. What's the reason for moving?

Comment (lotif) [reply tests/models/test_pca.py:12]: Actually, I take it back. This can potentially mess up other tests.  Is there an easy way to "unset" the seed? That way we can set them on startup and unset them on teardown.

Comment (yc7z) [reply tests/models/test_pca.py:12]: That was actually my reason for moving it under the functions that create the data matrix, because I thought doing this would make sure the seed is only set locally. But I just checked and you are right, calling it once make all the random functions that follow deterministic. I'm looking into possible ways to only set the seed temporarily

Comment (lotif) [reply tests/models/test_pca.py:12]: You can create an `unset_all_random_seeds` and set them to `None`, which will revert the change. Then you can make a pytest set up function that sets the seeds and a teardown function that unsets them (see [here](https://medium.com/@ramanish1992/pytest-setup-tear-down-configuration-c8197558a8e2) for an example)

**Comment by emersodb** [comment]
@yc7z and @lotif: Is this fix read to go in or do we still need some work on it?

**Comment by lotif** [comment]
@emersodb  I think setting the seed should be fixed but this was there before this PR so I'd approve without the fix if we have a clickup to fix it later.

File: tests/models/test_pca.py (hunk 1 above)
Comment (yc7z) [reply tests/models/test_pca.py:12]: Can you take a look at the latest changes? I think this should work now...

File: fl4health/utils/random.py [hunk 2]
```diff
@@ -22,3 +22,14 @@ def set_all_random_seeds(seed: Optional[int] = 42) -> None:
...
+
+def unset_all_random_seeds() -> None:
+    """
+    Set random seeds for Python random, NumPy, and PyTorch to None. Running this function would undo
+    the effects of set_all_random_seeds.
+    """
+    log(INFO, "Setting all random seeds to None.")
+    random.seed(None)
+    np.random.seed(None)
+    torch.manual_seed(None)
```
Comment (lotif) [inline fl4health/utils/random.py:35]: For pytorch, the correct call should be `torch.seed()`

File: tests/models/test_pca.py [hunk 3]
```diff
@@ -1,18 +1,30 @@
...
 
 
-def create_full_rank_data() -> Tensor:
+# For more inforamtion of this type hint,
+# see https://docs.python.org/3.8/library/typing.html#typing.Generator/
+@pytest.fixture
+def setup_random_seeds() -> Generator[None, None, None]:
```
Comment (lotif) [inline tests/models/test_pca.py:21]: That's a weird type hint :laughing:

File: tests/models/test_pca.py [hunk 4]
```diff
@@ -1,18 +1,30 @@
...
+    set_all_random_seeds(seed)
+    yield
+    unset_all_random_seeds()
+
+
+def create_full_rank_data(setup_random_seeds: Generator[None, None, None]) -> Tensor:
```
Comment (lotif) [inline tests/models/test_pca.py:27]: The param is not needed here, since this is not a test.

**Review by lotif — CHANGES_REQUESTED** [review]
Almost there, thanks for implementing it!

**Review by lotif — APPROVED** [review]
All good now, thanks for fixing!
